### ISSUE TITLE
Does not correctly narrow down `val: Any | None` by `val is not None`.

### ISSUE THREAD

**Issue by cainmagi** [microsoft/pyright#5508]
### **Describe the bug**
Using `is` expression to compare an `Optional` variable with a `None` will let the type checker narrow down the variable type. This technique works in most cases. However, it does not work properly if the variable is `Optional[Any]`, where the `is not None` branch narrows down the variable to `Any` correctly, while `is None` does not narrows the variable.

This is a further issue related to #4995, where Eric has explained the theoretical mechanism of the narrowing down. Here my issue is a little bit different from the example in the previous issue. In the following parts, I will only discuss the narrowing down for **only one variable**, which should reflect the core of the issue more properly.

At the same time, I think this issue may also be related to #5490. Hopefully, this can help the implementation of `is`-based type checking.

### **Codes and How to Reproduce**

Have tested this issue in the following 4 cases:
```python
from typing import Optional, Any
from typing_extensions import reveal_type


def test_int_is_none(val: Optional[int]):  # As expected.
    if val is None:
        reveal_type(val)  # Returned `None`, as expected.
    else:
        reveal_type(val)  # Returned `int` as expected.


def test_any_is_none(val: Optional[Any]):  # Abnormal.
    if val is None:
        reveal_type(val)  # Returned `Any | None`, however, expecting `None`.
    else:
        reveal_type(val)  # Returned `Any`, as expected.


def test_int_is_not_none(val: Optional[int]):  # As expected.
    if val is not None:
        reveal_type(val)  # Returned `int`, as expected.
    else:
        reveal_type(val)  # Returned `None` as expected.


def test_any_is_not_none(val: Optional[Any]):  # Abnormal.
    if val is not None:
        reveal_type(val)  # Returned `Any`, as expected.
    else:
        reveal_type(val)  # Returned `Any | None`, however, expecting `None`.
```

Besides, `Unknown` has the same issue.

```python
from typing_extensions import reveal_type

val = dict().get("any")
if val is None:
    reveal_type(val)  # Returned `Unknown | None`, expecting `None`.
else:
    reveal_type(val)  # Returned `Unknown`, as expected.
```

### **Expected behavior**

```python
from typing import Optional, Any
from typing_extensions import reveal_type


def test_any_is_none(val: Optional[Any]):  # Abnormal.
    if val is None:
        reveal_type(val)  # Should be annotated as `None`.
    else:
        reveal_type(val)  # Returned `Any`, as expected.


def test_any_is_not_none(val: Optional[Any]):  # Abnormal.
    if val is not None:
        reveal_type(val)  # Returned `Any`, as expected.
    else:
        reveal_type(val)  # Should be annotated as `None`.
```

### **VS Code extension or command-line**

* **Pylance** version: 2023.7.20
* **Pyright** version: 74d8f3c4

**Comment by erictraut** [comment]
In general, `Any` is not narrowed and is left "as is". However, there are a few exemptions to this rule that include `isinstance` and `issubclass` type guards and user-defined `TypeGuard`s. I think it's safe and reasonable to add the `is None` type guard pattern to this list of exemptions. Mypy appears to exempt this type guard as well.

I've updated the type guard logic and the documentation accordingly.

**Comment by erictraut** [comment]
Our compatibility tests indicated that some code bases rely on the fact that `Any` is not narrowed to `None` with the `is None` type guard in the positive (if) narrowing case. These same tests indicate that it is OK to narrow `Any | None` to `None`. I've adjusted the logic accordingly.

**Comment by cainmagi** [comment]
Thank you!

**Comment by erictraut** [comment]
Reopening. We don't close issues until the code has been released.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.318, which I just published. It will also be included in a future release of pylance.
